### ISSUE TITLE
operational callback seems to be called for each leaf even if it is returning all information on the first callback

### ISSUE THREAD

**Issue by bbergquist0930** [sysrepo/sysrepo#2675]
sysrepoctl - sysrepo YANG schema manipulation tool, compiled with libsysrepo v2.0.53 (SO v6.4.19)

I modified the `oper_data_pull_example.c` to allow multiple paths to be specified 
```
/**
 * @file oper_data_pull_example.c
 * @author Michal Vasko <mvasko@cesnet.cz>
 * @brief example of an application providing some operational data by a callback
 *
 * @copyright
 * Copyright (c) 2019 CESNET, z.s.p.o.
 *
 * This source code is licensed under BSD 3-Clause License (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://opensource.org/licenses/BSD-3-Clause
 */

#define _QNX_SOURCE /* sleep() */
#define _GNU_SOURCE

#include <inttypes.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <libyang/libyang.h>

#include "sysrepo.h"

volatile int exit_application = 0;

static int
dp_get_items_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_name, const char *xpath,
        const char *request_xpath, uint32_t request_id, struct lyd_node **parent, void *private_data)
{
    (void)session;
    (void)sub_id;
    (void)request_xpath;
    (void)request_id;
    (void)private_data;

    printf("\n\n ========== DATA FOR \"%s\" \"%s\" REQUESTED =======================\n\n", module_name, xpath);

    if (!strcmp(module_name, "examples") && strstr(xpath, "/examples:stats")) {
        lyd_new_path(NULL, sr_get_context(sr_session_get_connection(session)), "/examples:stats/counter", "852", 0, parent);
        lyd_new_path(*parent, NULL, "/examples:stats/counter2", "1052", 0, NULL);
    }

    return SR_ERR_OK;
}

static void
sigint_handler(int signum)
{
    (void)signum;

    exit_application = 1;
}

int
main(int argc, char **argv)
{
    sr_conn_ctx_t *connection = NULL;
    sr_session_ctx_t *session = NULL;
    sr_subscription_ctx_t *subscription = NULL;
    sr_subscr_options_t opt = SR_SUBSCR_DEFAULT;
    int rc = SR_ERR_OK;
    const char *mod_name;

    if (argc < 3) {
        printf("%s <module-to-provide-data-from> <path-to-provide> [<path-to-provide> ...]\n", argv[0]);
        return EXIT_FAILURE;
    }
    mod_name = argv[1];

    for (int i = 2; i < argc; i++) {
    	printf("Application will provide data \"%s\" of \"%s\".\n\n", argv[i], mod_name);
    }

    /* turn logging on */
    sr_log_stderr(SR_LL_WRN);

    /* connect to sysrepo */
    rc = sr_connect(0, &connection);
    if (rc != SR_ERR_OK) {
        goto cleanup;
    }

    /* start session */
    rc = sr_session_start(connection, SR_DS_RUNNING, &session);
    if (rc != SR_ERR_OK) {
        goto cleanup;
    }

    for (int i = 2; i < argc; i++) {
		/* subscribe for providing the operational data */
		rc = sr_oper_get_items_subscribe(session, mod_name, argv[i], dp_get_items_cb, NULL, opt, &subscription);
		if (rc != SR_ERR_OK) {
			goto cleanup;
		}
		opt = SR_SUBSCR_CTX_REUSE;
    }

    printf("\n\n ========== LISTENING FOR REQUESTS ==========\n\n");

    /* loop until ctrl-c is pressed / SIGINT is received */
    signal(SIGINT, sigint_handler);
    signal(SIGPIPE, SIG_IGN);
    while (!exit_application) {
        sleep(1000);
    }

    printf("Application exit requested, exiting.\n");

cleanup:
    sr_disconnect(connection);
    return rc ? EXIT_FAILURE : EXIT_SUCCESS;
}

```
I run this application and then issue the following 
```
brett@brett-Precision-7740:~/git/netconf/sysrepo$ sudo sysrepocfg -X -d operational -x /examples:stats
<stats xmlns="urn:examples">
  <counter>852</counter>
  <counter2>1052</counter2>
</stats>
brett@brett-Precision-7740:~/git/netconf/sysrepo$ 
```
The output from the application looks like 
```
brett@brett-Precision-7740:~/git/netconf/sysrepo$ sudo build/examples/oper_data_pull_example examples /examples:stats/counter /examples:stats/counter2
Application will provide data "/examples:stats/counter" of "examples".

Application will provide data "/examples:stats/counter2" of "examples".



 ========== LISTENING FOR REQUESTS ==========



 ========== DATA FOR "examples" "/examples:stats/counter" REQUESTED =======================



 ========== DATA FOR "examples" "/examples:stats/counter2" REQUESTED =======================

```
The callback was called multiple times.   If I run the application using 
```
brett@brett-Precision-7740:~/git/netconf/sysrepo$ sudo build/examples/oper_data_pull_example examples /examples:stats/counter /examples:stats
Application will provide data "/examples:stats/counter" of "examples".

Application will provide data "/examples:stats" of "examples".



 ========== LISTENING FOR REQUESTS ==========



 ========== DATA FOR "examples" "/examples:stats" REQUESTED =======================



 ========== DATA FOR "examples" "/examples:stats/counter" REQUESTED =======================


````
Then it is still being called twice.  I was under the impression that a callback could return extra/all of the data in which case sysrepo would use that data and not invoke the callback again.   I believe it used to work that way at least in Sysrepo 1.x.   I could go back and verify again if needed.

If being called multiple times, what is the sense of providing the extra data?   This can really impact performance if the implementation of the callback to acquire the data is expensive.

Is this behavior the expected behavior or am I doing something wrong with the callback registration or callback impementation?

**Comment by bbergquist0930** [comment]
I went back and verified with `sysrepoctl - sysrepo YANG schema manipulation tool, compiled with libsysrepo v1.4.58 (SO v5.5.0)
` and updated `oper_data_example` in the same way.   This is what I see 
```
brett@brett-Precision-7740:~/tmp/netopeer2/sysrepo/build$ ./sysrepocfg -X -d operational -x /examples:stats
<stats xmlns="urn:examples">
  <counter>852</counter>
  <counter2>1052</counter2>
</stats>
brett@brett-Precision-7740:~/tmp/netopeer2/sysrepo/build$ 
```
and
```
brett@brett-Precision-7740:~/tmp/netopeer2/sysrepo/build$ examples/oper_data_example examples /examples:stats
Application will provide data "/examples:stats" of "examples".



 ========== LISTENING FOR REQUESTS ==========



 ========== DATA FOR "examples" "/examples:stats" REQUESTED =======================


```

So in fact the sysrepo 1.x only calls the callbacks once.   I guess this is why when I upgrade to sysrepo 2.x, my client application started getting timeouts all over the place and everything felt much slower.

Do we know have to do our own caching of operational data in our application callbacks?   This seems like a very big issue with moving forward.

**Comment by jktjkt** [comment]
Your example for sysrepo v1.4.58 is using a single subscription for the enclosing container, and your CB provides two YANG leafs. That's a correct behavior. Your example for sysrepo v2 uses two active subscriptions, one for reach leaf. It's no surprise that both are called -- even if they just happen to execute the same function which happens to return two leafs. Why don't you subscribe for the enclosing container with sysrepo v2?

**Comment by bbergquist0930** [comment]
I try to be pretty thorough but I did make a mistake in the second run that I did not see until just now when I looked back after the comment.  I ran the example the second time with 
```
sudo build/examples/oper_data_pull_example examples /examples:stats/counter /examples:stats
```
when I meant to run it with 
```
sudo build/examples/oper_data_pull_example examples /examples:stats
```
I just did re-run with that and in fact the callback is called only once.    That is great that there is no regression.   I do have a separate question that is still related to this but will ask separately.

I apologize for this.
